ctapipe/reco/shower_processor.py

```diff
@@ -0,0 +1,176 @@
+            this is mutually exclusive with passing ``config``
+        """
+
+        super().__init__(config=config, parent=parent, **kwargs)
+        self.subarray = subarray
+        self.check_shower = ShowerQualityQuery(parent=self)
+        self.reconstructor = HillasReconstructor()
+
+    def reconstruct_shower(
+        self,
+        event,
+        default=DEFAULT_SHOWER_PARAMETERS,
+    ) -> ReconstructedGeometryContainer:
+        """Perform shower reconstruction.
+
+        Parameters
+        ----------
+        event : container
+            A ``ctapipe`` event container
+        default: container
+            The default 'ReconstructedGeometryContainer' which is
+            filled with NaNs.
+        Returns
+        -------
+        ReconstructedGeometryContainer:
+            direction in the sky with uncertainty,
+            core position on the ground with uncertainty,
+            h_max with uncertainty,
+            is_valid boolean for successfull reconstruction,
+            average intensity of the intensities used for reconstruction,
+            measure of algorithm success (if fit),
+            list of tel_ids used if stereo, or None if Mono
+        """
+
+        # Select only images which pass the shower quality criteria
+        hillas_dict = {
+            tel_id: dl1.parameters.hillas
+            for tel_id, dl1 in event.dl1.tel.items()
+            if all(self.check_shower(dl1.parameters))
+        }
+        self.log.debug(
+            "shower_criteria:\n %s", self.check_shower
+        )
+
+        # Reconstruct the shower only if all shower criteria are met
+        if len(hillas_dict) > 2:
+
+            array_pointing = SkyCoord(
+                az=event.pointing.array_azimuth,
+                alt=event.pointing.array_altitude,
+                frame=AltAz(),
+            )
+
+            telescopes_pointings = {
+                tel_id: SkyCoord(
+                    alt=event.pointing.tel[tel_id].altitude,
+                    az=event.pointing.tel[tel_id].azimuth,
+                    frame=AltAz(),
+                )
+                for tel_id in hillas_dict
+            }
+
+            result = self.reconstructor.predict(hillas_dict,
+                                                self.subarray,
+                                                array_pointing,
+                                                telescopes_pointings)
+
+            return result
+
+        else:
+            self.log.debug(
+                """Less than 2 images passed the quality cuts.
+                Returning default ReconstructedGeometryContainer container"""
+            )
+            return default
+
+    def process_shower_geometry(self, event: ArrayEventContainer):
+        """Record the reconstructed shower geometry into the ArrayEventContainer."""
+
+        shower_geometry = self.reconstruct_shower(event)
+
+        self.log.debug("shower geometry: %s", shower_geometry.as_dict(recursive=True))
```

Comment:
This will also already call `as_dict` for each event, not only when it actually is logged.

Reply:
how should it be printed? would a simpler

`self.log.debug("shower geometry:\n %s", shower_geometry)`

work?

Reply:
no, that wouldn't show the values, just the container definition. https://github.com/cta-observatory/ctapipe/issues/1627

Reply:
what would be a work-around for now?

Reply:
not logging here.

Reply:
done